File: BlotoutAnalytics/BlotoutAnalyticsAll/Events/BOAEventsManager.m
//
//  BOEventsManager.m
//  BlotoutAnalyticsTVOS
//
//  Created by ankuradhikari on 22/03/21.
//  Copyright © 2021 Blotout. All rights reserved.
//

#import "BOAEventsManager.h"
#import <BlotoutFoundation/BOFLogs.h>
#import "BOEventsOperationExecutor.h"
#import "BOADeveloperEvents.h"
#import "BOEventPostAPI.h"
#import <BlotoutFoundation/BOFLogs.h>
#import "BOAUtilities.h"

NSString *const BOAQueueKey = @"BOAQueue";
NSString *const kBOAQueueFilename = @"blotout.queue.plist";


@interface BOAEventsManager ()
@property (nonatomic, strong) NSMutableArray *queue;
@property (nonatomic, strong) id<BOAStorage> storage;
@property (nonatomic, strong) BlotoutAnalyticsConfiguration *configuration;
@property (nonatomic, strong) NSTimer *flushTimer;
@property (atomic, copy) NSDictionary *referrer;
@property (nonatomic, assign) UIBackgroundTaskIdentifier flushTaskID;
@property (nonatomic, readwrite) BOOL batchRequest;

@end

@implementation BOAEventsManager

- (id)initWithConfiguration:(BlotoutAnalyticsConfiguration *)configuration storage:(id<BOAStorage>)storage {
    
    if (self = [super init]) {
        self.configuration = configuration;
        self.storage = storage;
        
        self.flushTimer = [NSTimer timerWithTimeInterval:self.configuration.flushInterval
                                                  target:self
                                                selector:@selector(flush)
                                                userInfo:nil
                                                 repeats:YES];
        
        [NSRunLoop.mainRunLoop addTimer:self.flushTimer
                                forMode:NSDefaultRunLoopMode];
    }
    
    return self;
}

- (void)beginBackgroundTask
{
    @try {
        [self endBackgroundTask];
        
        [[BOEventsOperationExecutor sharedInstance] dispatchBackgroundTask:^{
            id<BOAApplicationProtocol> application = self.configuration.application;
            if (application) {
                self.flushTaskID = [application boa_beginBackgroundTaskWithName:@"BlotoutAnalytics_Background_Task"
                                                              expirationHandler:^{
                    [self endBackgroundTask];
                }];
            }
        }];
    } @catch (NSException *exception) {
        BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
    }
}

- (void)endBackgroundTask
{
    @try {
        [[BOEventsOperationExecutor sharedInstance] dispatchBackgroundTask:^{
            if (self.flushTaskID != UIBackgroundTaskInvalid) {
                id<BOAApplicationProtocol> application = self.configuration.application;
                if (application) {
                    [application boa_endBackgroundTask:self.flushTaskID];
                }
                
                self.flushTaskID = UIBackgroundTaskInvalid;
            }
        }];
    } @catch (NSException *exception) {
        BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
    }
}

- (void)capture:(BOACaptureModel *)payload
{
    @try {
        NSDictionary *event = [BOADeveloperEvents captureEvent:payload];
        if(event != nil) {
            [self enqueueEvent:@"capture" dictionary:event];
        }
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

-(void)capturePersonal:(BOACaptureModel *)payload isPHI:(BOOL)phiEvent {
    @try {
        NSDictionary *personalEvent = [BOADeveloperEvents capturePersonalEvent:payload isPHI:phiEvent];
        if(personalEvent != nil) {
            [self enqueueEvent:@"capturePersonal" dictionary:personalEvent];
        }
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)enqueueEvent:(NSString *)action dictionary:(NSDictionary *)payload {
    @try {
        [self queuePayload:payload];
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)queuePayload:(NSDictionary *)payload
{
    @try {
        payload = [BOAUtilities traverseJSON:payload];
        [self.queue addObject:payload];
        [self persistQueue];
        [self flushQueueByLength];
    }
    @catch (NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)flush
{
    [self flushWithMaxSize:0];
}

- (void)flushWithMaxSize:(NSUInteger)maxBatchSize
{
    @try {
        [[BOEventsOperationExecutor sharedInstance] dispatchEventsInBackground:^{
            if ([self.queue count] == 0) {
                BOFLogDebug(@"%@ No queued API calls to flush.", self);
                [self endBackgroundTask];
                return;
            }
            if (self.batchRequest) {
                BOFLogDebug(@"%@ API request already in progress, not flushing again.", self);
                return;
            }
            
            NSArray *batch = [NSArray arrayWithArray:self.queue];
            
            [self sendData:batch];
        }];
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)flushQueueByLength
{
    @try {
        [[BOEventsOperationExecutor sharedInstance] dispatchEventsInBackground:^{
            if (!self.batchRequest && [self.queue count] >= self.configuration.flushAt) {
                [self flush];
            }
        }];
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)sendData:(NSArray *)batch {
    @try {
        
        self.batchRequest = YES;
        [[BOEventsOperationExecutor sharedInstance] dispatchEventsInBackground:^{
            BOEventPostAPI *post = [[BOEventPostAPI alloc] init];
            NSDictionary *json =  [BOADeveloperEvents prepareServerPayload:batch];
            NSError *error = nil;
            NSData *data = [NSJSONSerialization dataWithJSONObject:json options:kNilOptions error:&error];
            
            [post postEventDataModel:data withAPICode:BOUrlEndPointEventDataPOST success:^(id  _Nonnull responseObject) {
                [self.queue removeObjectsInArray:batch];
                [self persistQueue];
                self.batchRequest = NO;
                [self endBackgroundTask];
            } failure:^(NSURLResponse * _Nonnull urlResponse, id  _Nonnull dataOrLocation, NSError * _Nonnull error) {
                self.batchRequest = NO;
                NSLog(@"%@",[error description]);
            }];
        }];
        
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
        self.batchRequest = NO;
    }
}

- (void)applicationDidEnterBackground
{
    @try {
        [self beginBackgroundTask];
        // We are gonna try to flush as much as we reasonably can when we enter background
        // since there is a chance that the user will never launch the app again.
        [self flush];
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (void)applicationWillTerminate
{
    @try {
        [[BOEventsOperationExecutor sharedInstance] dispatchInBackgroundAndWait:^{
            if (self.queue.count)
                [self persistQueue];
        }];
    } @catch(NSException *exception) {
        BOFLogDebug(@"%@", exception);
    }
}

- (NSMutableArray *)queue
{
    if (!_queue) {
#if TARGET_OS_TV
        _queue = [[self.storage arrayForKey:BOAQueueKey] ?: @[] mutableCopy];
#else
        _queue = [[self.storage arrayForKey:kBOAQueueFilename] ?: @[] mutableCopy];
#endif
    }
    
    return _queue;
}

- (void)persistQueue
{
#if TARGET_OS_TV
    [self.storage setArray:[self.queue copy] forKey:BOAQueueKey];
#else
    [self.storage setArray:[self.queue copy] forKey:kBOAQueueFilename];
#endif
}
@end

File: BlotoutAnalytics/Events/BOAEventsManager.swift
//
//  BOAEventsManager.swift
//  BlotoutAnalyticsSDK
//
//  Created by Poonam Tiwari on 22/03/22.
//

import Foundation
import UIKit

let BOAQueueKey = "BOAQueue"
let kBOAQueueFilename = "blotout.queue.plist"

class BOAEventsManager:NSObject {
    private var queue: [EventModel] = []
    private var storage: BOAStorage?
    private var configuration: BlotoutAnalyticsConfiguration?
    private var flushTimer: Timer?
    private var referrer: [AnyHashable : Any] = [:]
    private var flushTaskID: UIBackgroundTaskIdentifier?
    private var batchRequest = false
    private var postEventinProgress = false
    
//    required init?(coder aDecoder: NSCoder) {
//
//        super.init(coder: aDecoder)
//    }
    
    //TODO: check the method flow
    init(configuration: BlotoutAnalyticsConfiguration, storage: BOAStorage){
        
        super.init()
        //TODO: check this
        self.configuration = configuration
        self.storage = storage
        flushTimer = Timer(
            timeInterval: configuration.flushInterval,
            target: self,
            selector: #selector(flush),
            userInfo: nil,
            repeats: true)
        
        RunLoop.main.add(flushTimer!, forMode: .default)
        setQueueValue()
    }
    
    func beginBackgroundTask() {
        endBackgroundTask()

        BOEventsOperationExecutor.sharedInstance.dispatchBackgroundTask({ [self] in
                let application = configuration!.application
                if application == nil {
                    return
                }

                flushTaskID = application!.boa_beginBackgroundTask(
                    withName: "BlotoutAnalytics_Background_Task",
                    expirationHandler: { [self] in
                        endBackgroundTask()
                    })
            })
    }
    
    
    func endBackgroundTask() {
      
            BOEventsOperationExecutor.sharedInstance.dispatchBackgroundTask({ [self] in
                if flushTaskID == .invalid {
                    return
                }
                
                weak var application = configuration!.application
                if let application = application {
                    application.boa_endBackgroundTask(flushTaskID!)
                }
                
                flushTaskID = UIBackgroundTaskIdentifier.invalid
            })
       
    }
    
    func capture(_ payload: BOACaptureModel?) {
            let event = BOADeveloperEvents.captureEvent(payload)
            if event == nil {
                return
            }
            
       enqueueEvent("capture", eventModel: event)
    }
    
//    func capturePersonal(_ payload: BOACaptureModel?, isPHI phiEvent: Bool) {
//
//            let personalEvent = BOADeveloperEvents.capturePersonalEvent(payload, isPHI: phiEvent)
//            if personalEvent == nil {
//                return
//            }
//
//            enqueueEvent("capturePersonal", dictionary: personalEvent)
//    }
    
    func enqueueEvent(_ action: String?, eventModel payload: EventModel?) {
        
            queuePayload(payload)
    }
    
    
    /*
    func enqueueEvent(_ action: String?, dictionary payload: [AnyHashable : Any]?) {
        
            queuePayload(payload)
        
    }*/
    func queuePayload(_ payload: EventModel?) {
            //TODO: confirm later ,Maybe not needed
            //payload = BOAUtilities.traverseJSON(payload)
            if let payload = payload {
                    self.queue.insert(payload , at: 0)
            }
            persistQueue()
            flushQueueByLength()
    }
    
   /* func queuePayload(_ payload: [AnyHashable : Any]?) {
        var payload = payload

            //TODO: confirm later ,Maybe not needed
            //payload = BOAUtilities.traverseJSON(payload)
            if let payload = payload {
                    self.queue.insert(payload , at: 0)
            }
            persistQueue()
            flushQueueByLength()
        
    }*/
    
    @objc public func flush() {
        self.flushWithMaxSize(0)
    }
    
    func flushWithMaxSize(_ maxBatchSize: Int) {

            BOEventsOperationExecutor.sharedInstance.dispatchEvents(inBackground: { [self] in
                if queue.count == 0 {
                    BOFLogDebug(frmt: "%@ No queued API calls to flush.", args: self)
                    endBackgroundTask()
                    return
                }
                if batchRequest {
                    BOFLogDebug(frmt: "%@ API request already in progress, not flushing again.", args: self)
                    return
                }
                
                let batch = queue
                sendData(batch)
            })
    }
    
    
    func flushQueueByLength() {

            BOEventsOperationExecutor.sharedInstance.dispatchEvents(inBackground: { [self] in
                if !batchRequest && queue.count >= configuration!.flushAt {
                    flush()
                }
            })
    }
    
    
    func sendData(_ batch: [EventModel]) {
        batchRequest = true
        postEventinProgress = true
        BOEventsOperationExecutor.sharedInstance.dispatchEvents(inBackground: { [self] in
            let post = BOEventPostAPI()
            let json:[AnyHashable:Any]? = BOADeveloperEvents.prepareServerPayload(events: batch)
            var error: Error? = nil
            do
            {
                let data: Data? = try JSONSerialization.data(withJSONObject: json, options: [])
                post.postEventData(data) { success in
                    //TODO: need to fix this
                    self.queue = self.queue.filter({ !batch.contains($0) })
                    self.persistQueue()
                    self.batchRequest = false
                    self.postEventinProgress = false
                    self.endBackgroundTask()
                    print("success")
                    
                } failure: { error in
                    self.postEventinProgress = false
                    print("failed no network")
                    //if this error is network error, listen for changes in network & retry again later
                    BOAReachability.sharedInstance.netStatusChangeHandler = {
                        //need to make api call again here, if its connected now & we have pending items
                        print("errored due to network, make api call again")
                        let batch = self.queue
                        if batch.count > 0 && (!self.postEventinProgress){
                            self.sendData(batch)
                            self.postEventinProgress = true
                        }
                    }
                    
                    self.batchRequest = false
                    BOFLogDebug(frmt: "%@", args: error?.localizedDescription as! CVarArg)
                }
            }
            catch
            {
                batchRequest = false
                BOFLogDebug(frmt: "%@", args: error.localizedDescription as! CVarArg)
            }
        })
    }


    
    func applicationDidEnterBackground() {
            beginBackgroundTask()
            // We are gonna try to flush as much as we reasonably can when we enter background
            // since there is a chance that the user will never launch the app again.
            flush()
    }
    
    func applicationWillTerminate() {
            BOEventsOperationExecutor.sharedInstance.dispatch(inBackgroundAndWait: { [self] in
                if queue.count > 0 {
                    persistQueue()
                }
            })

    }
    
    //TODO: have changed name,need to verify code

    func setQueueValue()
    {
#if os(tvOS)
            self.queue = (storage.array(forKey: BOAQueueKey) ?? [])
#else
        if let newQueue = storage?.arrayForKey(kBOAQueueFilename)
        {
            self.queue = newQueue //as! [EventModel]
        }
#endif
    }
    
    func persistQueue() {
#if os(tvOS)
        storage.set(queue, forKey: BOAQueueKey)
#else
        storage!.setArray(queue, forKey: kBOAQueueFilename)
#endif
    }
    
}
